File: apps/backend/src/Infrastructure/Schools/SchoolService.cs
using Kcow.Application.Interfaces;
using Kcow.Application.Schools;
using Kcow.Domain.Entities;
using Microsoft.Extensions.Logging;

namespace Kcow.Infrastructure.Schools;

/// <summary>
/// Implementation of school management service using Dapper repositories.
/// </summary>
public class SchoolService : ISchoolService
{
    private readonly ISchoolRepository _schoolRepository;
    private readonly ILogger<SchoolService> _logger;

    public SchoolService(ISchoolRepository schoolRepository, ILogger<SchoolService> logger)
    {
        _schoolRepository = schoolRepository;
        _logger = logger;
    }

    /// <summary>
    /// Gets all active schools.
    /// </summary>
    public async Task<List<SchoolDto>> GetAllAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            var schools = (await _schoolRepository.GetActiveAsync(cancellationToken))
                .OrderBy(s => s.Name)
                .ToList();

            _logger.LogInformation("Retrieved {Count} active schools", schools.Count);

            return schools.Select(MapToDto).ToList();
        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Error retrieving all schools");
            throw;
        }
    }

    /// <summary>
    /// Gets a school by ID.
    /// </summary>
    public async Task<SchoolDto?> GetByIdAsync(int id, CancellationToken cancellationToken = default)
    {
        var school = await _schoolRepository.GetByIdAsync(id, cancellationToken);

        if (school == null || !school.IsActive)
        {
            _logger.LogWarning("School with ID {SchoolId} not found", id);
            return null;
        }

        _logger.LogInformation("Retrieved school with ID {SchoolId}", id);
        return MapToDto(school);
    }

    /// <summary>
    /// Creates a new school.
    /// </summary>
    public async Task<SchoolDto> CreateAsync(CreateSchoolRequest request, CancellationToken cancellationToken = default)
    {
        var school = new School
        {
            Name = request.Name,
            ShortName = request.ShortName,
            SchoolDescription = request.SchoolDescription,
            TruckId = request.TruckId,
            Price = request.Price,
            FeeDescription = request.FeeDescription,
            Formula = request.Formula,
            VisitDay = request.VisitDay,
            VisitSequence = request.VisitSequence,
            ContactPerson = request.ContactPerson,
            ContactCell = request.ContactCell,
            Phone = request.Phone,
            Telephone = request.Telephone,
            Fax = request.Fax,
            Email = request.Email,
            CircularsEmail = request.CircularsEmail,
            Address = request.Address,
            Address2 = request.Address2,
            Headmaster = request.Headmaster,
            HeadmasterCell = request.HeadmasterCell,
            IsActive = true,
            Language = request.Language,
            PrintInvoice = request.PrintInvoice,
            ImportFlag = request.ImportFlag,
            Afterschool1Name = request.Afterschool1Name,
            Afterschool1Contact = request.Afterschool1Contact,
            Afterschool2Name = request.Afterschool2Name,
            Afterschool2Contact = request.Afterschool2Contact,
            SchedulingNotes = request.SchedulingNotes,
            MoneyMessage = request.MoneyMessage,
            SafeNotes = request.SafeNotes,
            WebPage = request.WebPage,
            Omsendbriewe = request.Omsendbriewe,
            KcowWebPageLink = request.KcowWebPageLink,
            CreatedAt = DateTime.UtcNow
        };

        var id = await _schoolRepository.CreateAsync(school, cancellationToken);
        school.Id = id; // Set the ID returned by repository

        _logger.LogInformation("Created school with ID {SchoolId} and name '{SchoolName}'",
            school.Id, school.Name);

        return MapToDto(school);
    }

    /// <summary>
    /// Updates an existing school.
    /// </summary>
    public async Task<SchoolDto?> UpdateAsync(int id, UpdateSchoolRequest request, CancellationToken cancellationToken = default)
    {
        var school = await _schoolRepository.GetByIdAsync(id, cancellationToken);

        if (school == null || !school.IsActive)
        {
            _logger.LogWarning("Cannot update: School with ID {SchoolId} not found", id);
            return null;
        }

        school.Name = request.Name;
        school.ShortName = request.ShortName;
        school.SchoolDescription = request.SchoolDescription;
        school.TruckId = request.TruckId;
        school.Price = request.Price;
        school.FeeDescription = request.FeeDescription;
        school.Formula = request.Formula;
        school.VisitDay = request.VisitDay;
        school.VisitSequence = request.VisitSequence;
        school.ContactPerson = request.ContactPerson;
        school.ContactCell = request.ContactCell;
        school.Phone = request.Phone;
        school.Telephone = request.Telephone;
        school.Fax = request.Fax;
        school.Email = request.Email;
        school.CircularsEmail = request.CircularsEmail;
        school.Address = request.Address;
        school.Address2 = request.Address2;
        school.Headmaster = request.Headmaster;
        school.HeadmasterCell = request.HeadmasterCell;
        school.Language = request.Language;
        school.PrintInvoice = request.PrintInvoice;
        school.ImportFlag = request.ImportFlag;
        school.Afterschool1Name = request.Afterschool1Name;
        school.Afterschool1Contact = request.Afterschool1Contact;
        school.Afterschool2Name = request.Afterschool2Name;
        school.Afterschool2Contact = request.Afterschool2Contact;
        school.SchedulingNotes = request.SchedulingNotes;
        school.MoneyMessage = request.MoneyMessage;
        school.SafeNotes = request.SafeNotes;
        school.WebPage = request.WebPage;
        school.Omsendbriewe = request.Omsendbriewe;
        school.KcowWebPageLink = request.KcowWebPageLink;
        school.UpdatedAt = DateTime.UtcNow;

        await _schoolRepository.UpdateAsync(school, cancellationToken);

        _logger.LogInformation("Updated school with ID {SchoolId}", id);

        return MapToDto(school);
    }

    /// <summary>
    /// Archives (soft-deletes) a school.
    /// </summary>
    public async Task<bool> ArchiveAsync(int id, CancellationToken cancellationToken = default)
    {
        var school = await _schoolRepository.GetByIdAsync(id, cancellationToken);

        if (school == null || !school.IsActive)
        {
            _logger.LogWarning("Cannot archive: School with ID {SchoolId} not found", id);
            return false;
        }

        school.IsActive = false;
        school.UpdatedAt = DateTime.UtcNow;

        await _schoolRepository.UpdateAsync(school, cancellationToken);

        _logger.LogInformation("Archived school with ID {SchoolId}", id);
        return true;
    }

    /// <summary>
    /// Maps a School entity to a SchoolDto.
    /// </summary>
    private static SchoolDto MapToDto(School school)
    {
        return new SchoolDto
        {
            Id = school.Id,
            Name = school.Name,
            ShortName = school.ShortName,
            SchoolDescription = school.SchoolDescription,
            TruckId = school.TruckId,
            Price = school.Price,
            FeeDescription = school.FeeDescription,
            Formula = school.Formula,
            VisitDay = school.VisitDay,
            VisitSequence = school.VisitSequence,
            ContactPerson = school.ContactPerson,
            ContactCell = school.ContactCell,
            Phone = school.Phone,
            Telephone = school.Telephone,
            Fax = school.Fax,
            Email = school.Email,
            CircularsEmail = school.CircularsEmail,
            Address = school.Address,
            Address2 = school.Address2,
            Headmaster = school.Headmaster,
            HeadmasterCell = school.HeadmasterCell,
            IsActive = school.IsActive,
            Language = school.Language,
            PrintInvoice = school.PrintInvoice,
            ImportFlag = school.ImportFlag,
            Afterschool1Name = school.Afterschool1Name,
            Afterschool1Contact = school.Afterschool1Contact,
            Afterschool2Name = school.Afterschool2Name,
            Afterschool2Contact = school.Afterschool2Contact,
            SchedulingNotes = school.SchedulingNotes,
            MoneyMessage = school.MoneyMessage,
            SafeNotes = school.SafeNotes,
            WebPage = school.WebPage,
            Omsendbriewe = school.Omsendbriewe,
            KcowWebPageLink = school.KcowWebPageLink,
            LegacyId = school.LegacyId,
            CreatedAt = school.CreatedAt,
            UpdatedAt = school.UpdatedAt
        };
    }
}

File: apps/frontend/src/app/core/services/school.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { Observable, tap, catchError, finalize, throwError, map } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface BillingSettings {
    defaultSessionRate?: number;
    billingCycle?: 'Monthly' | 'Termly';
    billingNotes?: string;
}

export interface School {
    id: number;
    name: string;
    shortName?: string;
    truckId?: number;
    price?: number;
    feeDescription?: string;
    formula?: number;
    visitDay?: string;
    visitSequence?: string;
    contactPerson?: string;
    contactCell?: string;
    phone?: string;
    telephone?: string;
    fax?: string;
    email?: string;
    circularsEmail?: string;
    address?: string;
    address2?: string;
    headmaster?: string;
    headmasterCell?: string;
    isActive: boolean;
    language?: string;
    printInvoice: boolean;
    importFlag: boolean;
    afterschool1Name?: string;
    afterschool1Contact?: string;
    afterschool2Name?: string;
    afterschool2Contact?: string;
    schedulingNotes?: string;
    moneyMessage?: string;
    safeNotes?: string;
    webPage?: string;
    kcowWebPageLink?: string;
    billingSettings?: BillingSettings;
}

export interface CreateSchoolRequest {
    name: string;
    shortName?: string;
    truckId?: number;
    price?: number;
    feeDescription?: string;
    formula?: number;
    visitDay?: string;
    visitSequence?: string;
    contactPerson?: string;
    contactCell?: string;
    phone?: string;
    telephone?: string;
    fax?: string;
    email?: string;
    circularsEmail?: string;
    address?: string;
    address2?: string;
    headmaster?: string;
    headmasterCell?: string;
    language?: string;
    printInvoice: boolean;
    importFlag: boolean;
    afterschool1Name?: string;
    afterschool1Contact?: string;
    afterschool2Name?: string;
    afterschool2Contact?: string;
    schedulingNotes?: string;
    moneyMessage?: string;
    safeNotes?: string;
    webPage?: string;
    kcowWebPageLink?: string;
    billingSettings?: BillingSettings;
}

export interface UpdateSchoolRequest {
    name: string;
    shortName?: string;
    truckId?: number;
    price?: number;
    feeDescription?: string;
    formula?: number;
    visitDay?: string;
    visitSequence?: string;
    contactPerson?: string;
    contactCell?: string;
    phone?: string;
    telephone?: string;
    fax?: string;
    email?: string;
    circularsEmail?: string;
    address?: string;
    address2?: string;
    headmaster?: string;
    headmasterCell?: string;
    isActive: boolean;
    language?: string;
    printInvoice: boolean;
    importFlag: boolean;
    afterschool1Name?: string;
    afterschool1Contact?: string;
    afterschool2Name?: string;
    afterschool2Contact?: string;
    schedulingNotes?: string;
    moneyMessage?: string;
    safeNotes?: string;
    webPage?: string;
    kcowWebPageLink?: string;
    billingSettings?: BillingSettings;
}

export interface ProblemDetails {
    type?: string;
    title?: string;
    status?: number;
    detail?: string;
    instance?: string;
    [key: string]: unknown;
}

@Injectable({
    providedIn: 'root',
})
export class SchoolService {
    private http = inject(HttpClient);
    private readonly apiUrl = `${environment.apiUrl}/schools`;

    // Signals for state
    private schoolsSignal = signal<School[]>([]);
    private isLoadingSignal = signal<boolean>(false);
    private errorSignal = signal<ProblemDetails | null>(null);

    // Read-only signals
    readonly schools = computed(() => this.schoolsSignal());
    readonly isLoading = computed(() => this.isLoadingSignal());
    readonly error = computed(() => this.errorSignal());

    /**
     * Fetch all schools from the backend
     */
    getSchools(): Observable<School[]> {
        this.isLoadingSignal.set(true);
        this.errorSignal.set(null);

        return this.http.get<School[]>(this.apiUrl, { withCredentials: true }).pipe(
            tap((schools) => {
                this.schoolsSignal.set(schools);
            }),
            catchError((error: HttpErrorResponse) => {
                // Return an observable that emits the mapped error
                return this.handleError(error);
            }),
            // Use tap only if there was an error to update the local error signal
            // This is safer than the nested tap in catchError
            tap({
                error: (err) => this.errorSignal.set(err)
            }),
            finalize(() => {
                this.isLoadingSignal.set(false);
            })
        );
    }

    /**
     * Get a single school by ID
     */
    getSchoolById(id: number): Observable<School> {
        return this.http.get<School>(`${this.apiUrl}/${id}`, { withCredentials: true }).pipe(
            catchError((error: HttpErrorResponse) => {
                return this.handleError(error);
            })
        );
    }

    /**
     * Create a new school
     */
    createSchool(school: CreateSchoolRequest): Observable<School> {
        return this.http.post<School>(this.apiUrl, school, { withCredentials: true }).pipe(
            tap((newSchool) => {
                // Update the schools signal with the new school
                this.schoolsSignal.update((schools) => [...schools, newSchool]);
            }),
            catchError((error: HttpErrorResponse) => {
                return this.handleError(error);
            })
        );
    }

    /**
     * Update an existing school
     */
    updateSchool(id: number, school: UpdateSchoolRequest): Observable<School> {
        return this.http.put<School>(`${this.apiUrl}/${id}`, school, { withCredentials: true }).pipe(
            tap((updatedSchool) => {
                // Update the schools signal with the updated school
                this.schoolsSignal.update((schools) =>
                    schools.map((s) => (s.id === id ? updatedSchool : s))
                );
            }),
            catchError((error: HttpErrorResponse) => {
                return this.handleError(error);
            })
        );
    }

    /**
     * Deactivate a school (soft delete - sets isActive to false)
     */
    deactivateSchool(id: number, school: School): Observable<School> {
        const { id: _, ...rest } = school;
        return this.updateSchool(id, {
            ...rest,
            isActive: false,
        });
    }

    /**
     * Reactivate a school (sets isActive back to true)
     */
    reactivateSchool(id: number, school: School): Observable<School> {
        const { id: _, ...rest } = school;
        return this.updateSchool(id, {
            ...rest,
            isActive: true,
        });
    }

    /**
     * Get only active schools (for dropdowns/selections)
     */
    getActiveSchools(): Observable<School[]> {
        return this.http.get<School[]>(this.apiUrl, { withCredentials: true }).pipe(
            map((schools: School[]) => schools.filter((s: School) => s.isActive)),
            catchError((error: HttpErrorResponse) => this.handleError(error))
        );
    }

    /**
     * Handle HTTP errors and map to ProblemDetails
     */
    private handleError(error: HttpErrorResponse): Observable<never> {
        let problemDetails: ProblemDetails = {
            title: 'An error occurred',
            detail: 'An unexpected error occurred',
            status: error.status,
        };

        if (error.error && typeof error.error === 'object') {
            problemDetails = { ...problemDetails, ...error.error };
        } else if (typeof error.error === 'string') {
            problemDetails.detail = error.error;
        }

        // Ensure title is present for standard ProblemDetails consistency
        if (!problemDetails.title && problemDetails.status) {
            problemDetails.title = `Error ${problemDetails.status}`;
        }

        return throwError(() => problemDetails);
    }
}
